runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,164 @@
+local M = {}
+local util = require "vim.lsp.util"
+local highlight = require "vim.highlight"
+local ns = vim.api.nvim_create_namespace("lsp-semantic-tokens")
+
+local semantic_tokens = {}
+
+local function modifiers_from_number(x, modifiers_table)
+  local function get_bit(n, k)
+    -- Based on/from https://stackoverflow.com/a/26230537
+    -- (n >> k) & 1
+    return _G.bit.band(_G.bit.rshift(n, k), 1)
+  end
+
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+M.token_map = {
+  type = "LspType",
+  class = "LspClass",
+  enum = "LspEnum",
+  interface = "LspInterface",
+  struct = "LspStruct",
+  typeParameter = "LspTypeParameter",
+  parameter = "LspParameter",
+  variable = "LspVariable",
+  property = "LspProperty",
+  enumMember = "LspEnumMember",
+  event = "LspEvent",
+  ["function"] = "LspFunction",
+  method = "LspMethod",
+  macro = "LspMacro",
+  keyword = "LspKeyword",
+  modifier = "LspModifier",
+  comment = "LspComment",
+  string = "LspString",
+  number = "LspNumber",
+  regexp = "LspRegexp",
+  operator = "LspOperator"
+}
+
+M.modifiers_map = {
+  declaration = "LspDeclaration",
+  definition = "LspDefinition",
+  readonly = "LspReadonly",
+  static = "LspStatic",
+  deprecated = "LspDeprecated",
+  abstract = "LspAbstract",
+  async = "LspAsync",
+  modification = "LspModification",
+  documentation = "LspDocumentation",
+  defaultLibrary = "LspDefaultLibrary"
+}
+
+---@private
+local function make_cache(base_map)
+  return setmetatable({}, { __index = function(_, ft)
+    return setmetatable({}, { __index = function(_, key)
+      local hl = base_map[key]
+      if hl then
+        return highlight._link_default_highlight_once(ft..hl, hl)
+      end
+      local id = vim.api.nvim_get_hl_id_by_name(hl)
+
+      rawset(table, key, id)
+      return id
+    end})
+  end})
+end
+
+local token_cache = make_cache(M.token_map)
+local modifiers_cache = make_cache(M.modifiers_map)
+
+---@private
+local function highlight(buf, token, hl)
+  vim.highlight.range(buf, ns, hl, {token.line, token.start_char}, {token.line, token.start_char + token.length})
```

Comment:
TODO: we don't lock the text after making our request, when getting the response the document might have actually changed. so the ranges, described here might be invalid.

Reply:
use changed_tick